File: vuln-sandbox/php/cwe-798/hardcoded-secret-safe.php
<?php
// CWE-798: Hardcoded Credentials - Safe Pattern
// Rule: php-hardcoded-secret
// Expected: TrueNegative

// SAFE: Credentials loaded from environment variables
$password = getenv('DB_PASSWORD');

// SAFE: Using $_ENV superglobal
$api_key = $_ENV['API_KEY'];

// SAFE: Using a configuration file outside web root
$config = parse_ini_file('/etc/app/config.ini');
$secret_key = $config['secret_key'];

class DatabaseConfig {
    // SAFE: Reading from environment
    private $token;

    public function __construct() {
        $this->token = getenv('AUTH_TOKEN');
    }
}
?>

File: vuln-sandbox/ruby/cwe-798/hardcoded-secret-safe.rb
# CWE-798: Hardcoded Credentials - Safe Pattern
# Rule: ruby-hardcoded-secret
# Expected: TrueNegative

# SAFE: Credentials loaded from environment variables
api_key = ENV["API_KEY"]
secret_key = ENV["SECRET_KEY"]

class PaymentService
  # SAFE: Using Rails credentials or environment variables
  def initialize
    @api_key = Rails.application.credentials.payment_api_key
    @password = ENV.fetch("DB_PASSWORD")
  end
end
